### ISSUE TITLE
EF Core issue when creating model using model builder

### ISSUE THREAD

**Issue by Luigi6821** [dotnet/efcore#23600]
Hi,
I am experiencing an issue while building model using EF Core (EF Core 5.0). 
Console application (NET5) fails when trying to SaveChanges: "AMOS_ADDRESS is not mapped to a Table"
as reported in the attached screenshot.
Please can you help me in understanding what I am doing wrong?
Thanks

           var loggerFactory = LoggerFactory.Create(builder => {
                builder.AddFilter("Microsoft", LogLevel.Debug)
                       .AddFilter("System", LogLevel.Debug)
                       .AddFilter("ConsoleApp4.Program", LogLevel.Debug)
                       .AddConsole();
            });
            var optionBuilder = new DbContextOptionsBuilder();
            optionBuilder.UseLazyLoadingProxies(true);
            optionBuilder.UseLoggerFactory(loggerFactory);

            var modelBuilder = new ModelBuilder();
            modelBuilder.HasDefaultSchema("AMOS");
            modelBuilder.Entity<AMOS_ADDRESS>(b => b.ToTable("ADDRESS"));
            modelBuilder.Entity<AMOS_ADDRESS>().Property(e => e.CODE).HasColumnName("CODE");
            modelBuilder.Entity<AMOS_ADDRESS>().Property(e => e.ALPHACODE).HasColumnName("ALPHACODE");
            modelBuilder.Entity<AMOS_ADDRESS>().Property(e => e.ADDRESSID).HasColumnName("ADDRESSID");
            modelBuilder.Entity<AMOS_ADDRESS>().HasKey(e => e.ADDRESSID);

            modelBuilder.FinalizeModel();
            optionBuilder.UseModel(modelBuilder.Model);

            //optionBuilder.UseInMemoryDatabase(databaseName: "myDb");
            var connection = Microsoft.Data.SqlClient.SqlClientFactory.Instance.CreateConnection();
            connection.ConnectionString = "Data source=localhost;Initial catalog=ABS10.0.30;User ID=amos;Password=voyager";
            connection.Open(); // Works
            // To prove that table exists and connection works
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "SELECT COUNT(1) from AMOS.ADDRESS";
                var count = (int)command.ExecuteScalar();
                Console.WriteLine($"Address count {count}");
            }
            optionBuilder.UseSqlServer(connection);

            var dbContext = new myDbContext(optionBuilder.Options);
            dbContext.Addresses.Add(new AMOS_ADDRESS() { CODE = "001", ADDRESSID = 1, ALPHACODE = "A001" });
            dbContext.Addresses.Add(new AMOS_ADDRESS() { CODE = "002", ADDRESSID = 2, ALPHACODE = "A002" });
            dbContext.Addresses.Add(new AMOS_ADDRESS() { CODE = "003", ADDRESSID = 3, ALPHACODE = "A003" });
            dbContext.Addresses.Add(new AMOS_ADDRESS() { CODE = "004", ADDRESSID = 4, ALPHACODE = "A004" });
            dbContext.SaveChanges(); // FAILS AMOS_ADDRESS is not mapped to a table

            var addresses = dbContext.Addresses;
            Console.WriteLine(addresses.CountAsync().Result); // Fails AMOS_ADDRESS is not mapped to a table

![image](https://user-images.githubusercontent.com/53300194/101356912-07fb6600-3899-11eb-84aa-15750d3450d3.png)

**Comment by roji** [comment]
Can you provide a bit more context on why you're creating your own ModelBuilder in this way? Normal EF Core usage usually depends on conventions which perform additional configuration (see the warning on the [ModelBuilder API docs](https://docs.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.modelbuilder.-ctor?view=efcore-5.0#Microsoft_EntityFrameworkCore_ModelBuilder__ctor)). Any reason you're not building your model in your DbContext's OnModelCreating, [as in the docs](https://docs.microsoft.com/en-us/ef/core/modeling/)?

**Comment by Luigi6821** [comment]
Hi @roji,
This code is a very simple "extract" of the final solution. The final solution is a web-API project where EF is used for managing
data (classes are more than 300). In this final solution the model is built only the first time (there is a framework that scans classes and apply configuration based on property attributes), cached and used in any subsequent calls.
I remember that I used it in previous EF core releases and it worked...

Regards
Luigi

**Comment by roji** [comment]
@Luigi6821 when the standard mechanism for building the model is used (i.e. via DbContext's OnModelCreating), the model is also built only the first time and then cached - I don't think what you're doing helps performance in any way.

It's also improbable the the code above worked as-is in previous releases - the model builder has no conventions whatsoever configured, but these are generally necessary for proper model building; note the other constructor of ModelBuilder, which accepts a ConventionSet. Setting up the proper ConventionSet is possible (with conventions from your specific provider, from relational and from core), but this really isn't the recommended way to work with EF Core.

May I suggest trying out the standard mechanism (via OnModelCreating), and seeing whether that works for you?

**Comment by smitpatel** [comment]
Possible cause of error - Relational model.
This may be working in 3.1 release. User is creating a conventionless model builder which excludes the convention to build the relational model. A lot of our stack assumes that relational model will be available.

**Comment by Luigi6821** [comment]
@roji 
The standard way (via OnModelCreating) works and model is cached as you correctly stated. 
Defining the following

var modelBuilder = new ModelBuilder(SqlServerConventionSetBuilder.Build());

works as expected also using external model definition.

A new one if possible...
Suppose a model like the following:

public abstract class AMOS_ADDRESS
{ 
   public decimal ADDRESSID { get; set; }
   public string CODE { get; set; }
   public string ALPHACODE { get; set; }
 }
 public class ADDRESS : AMOS_ADDRESS
 {
      public string ADDRESS1 { get; set; }
  }

And a model definition code as follows:

 modelBuilder.HasDefaultSchema("AMOS");
 modelBuilder.Entity<AMOS_ADDRESS>(b => b.ToTable("ADDRESS"));
 modelBuilder.Entity<AMOS_ADDRESS>().HasKey(e => e.ADDRESSID);
 modelBuilder.Entity<AMOS_ADDRESS>().Property(e => e.CODE).HasColumnName("CODE");
 modelBuilder.Entity<AMOS_ADDRESS>().Property(e => e.ALPHACODE).HasColumnName("ALPHACODE");
 modelBuilder.Entity<AMOS_ADDRESS>().Property(e => e.ADDRESSID).HasColumnName("ADDRESSID");
 
 modelBuilder.Entity<ADDRESS>(b => b.ToTable("ADDRESS"));
 modelBuilder.Entity<ADDRESS>().Property(e => e.ADDRESS1).HasColumnName("ADDRESS1");

 While EF supports it EF Core adds Discriminator column when using both methods

Context.Set<AMOS_ADDRESS>() or Context.Set<ADDRESS>()

It is an expected behavior or it will be changed in future?
Thanks

**Comment by roji** [comment]
@Luigi6821 yes, the discriminator column is added by-design, and AFAIK has been the EF Core behavior since version 1.0 (I also think EF6 did the same) - you probably want to take a look at our [docs on inheritance](https://docs.microsoft.com/en-us/ef/core/modeling/inheritance). Without the discriminator, EF cannot know which entity type is stored in a given database row (is it ADDRESS, or another subclass of AMOS_ADDRESS?). An alternative is to use TPT instead of TPH.

Leaving open in case we want to discuss @smitpatel's comment above.

**Comment by Luigi6821** [comment]
@roji 
I believe that EF Core should work as EF (NET Framework) which do not include discriminator if only one concrete class per abstract.
In fact the below code (using NET Framework) works without problem (model is same as EF core example)

    class Program
    {
        static void Main(string[] args)
        {
            var loggerFactory = LoggerFactory.Create(builder => {
                builder.AddFilter("Microsoft", LogLevel.Debug)
                       .AddFilter("System", LogLevel.Debug)
                       .AddFilter("ConsoleApp4.Program", LogLevel.Debug)
                       .AddConsole();
            });

            var modelBuilder = new DbModelBuilder();
            modelBuilder.HasDefaultSchema("AMOS");

            modelBuilder.Entity<AMOS_ADDRESS>().ToTable("ADDRESS", "AMOS");
            modelBuilder.Entity<AMOS_ADDRESS>().HasKey(e => e.ADDRESSID);

            modelBuilder.Entity<AMOS_ADDRESS>().Property(e => e.CODE).HasColumnName("CODE");
            modelBuilder.Entity<AMOS_ADDRESS>().Property(e => e.ALPHACODE).HasColumnName("ALPHACODE");
            modelBuilder.Entity<AMOS_ADDRESS>().Property(e => e.ADDRESSID).HasColumnName("ADDRESSID");

            modelBuilder.Entity<ADDRESS>().ToTable("ADDRESS", "AMOS");
            modelBuilder.Entity<ADDRESS>().Property(e => e.ADDRESS1).HasColumnName("ADDRESS1");

            var connection = System.Data.SqlClient.SqlClientFactory.Instance.CreateConnection();
            connection.ConnectionString = "Data source=localhost;Initial catalog=ABS10.0.30;User ID=amos;Password=voyager";
            connection.Open(); // Works

            var builtModel = modelBuilder.Build(connection).Compile();
            using (var dbContext = new DbContext(connection, builtModel, true))
            {
                var addressesBase = dbContext.Set<AMOS_ADDRESS>().ToArrayAsync().Result;
                var addressesDerived = dbContext.Set<ADDRESS>().ToArrayAsync().Result;
                Console.WriteLine("Press key to end");
                Console.ReadKey();
            }
        }

        public abstract class AMOS_ADDRESS
        {
            public decimal ADDRESSID { get; set; }

            public string CODE { get; set; }

            public string ALPHACODE { get; set; }
        }

        public class ADDRESS : AMOS_ADDRESS
        {
            public string ADDRESS1 { get; set; }
        }
    }

Please can you give me hints on how I can do same with EF Core?
Thanks

**Comment by roji** [comment]
Simply do all your model configuration on ADDRESS - don't do `modelBuilder.Entity<AMOS_ADDRESS>().ToTable(...)`. If you don't explicitly map/configure the base class, EF will create a simple, single table without a discriminator.

**Comment by Luigi6821** [comment]
@roji 
Yes it works but I can't do 

dbContext.Set<**AMOS_ADDRESS**>()

Because of **"this type is not included in the model for the context"**

**Comment by roji** [comment]
> dbContext.Set<AMOS_ADDRESS>()
> [...]
>  I always get ADDRESS instances.

I think you said AMOS_ADDRESS is abstract, no? So why would you need to have `dbContext.Set<AMOS_ADDRESS>()`, and doesn't it make sense for you to always get ADDRESS instances (since it's the only concrete type)?

Basically, if there's more than one possible concrete type in the hierarchy, a discriminator is needed to distinguish between them in the database. Otherwise, you can just map the concrete class directly (regardless of whether it inherits from an abstract base class or not).

**Comment by Luigi6821** [comment]
> I think you said AMOS_ADDRESS is abstract, no? So why would you need to have `dbContext.Set<AMOS_ADDRESS>()`, and doesn't it make sense for you to always get ADDRESS instances (since it's the only concrete type)?

Yes it makes sense to have always ADDRESS as type even if I use dbContext.Set<AMOS_ADDRESS>().
I explain what I am doing if you have patience to read it :-).
The basic idea is to offer to our (my company) customers possibility to customize our base classes (they consist of more than 400 tables - maybe not all but a bunch of them) adding new columns for usage in their own customizations.
Our SDK is distributed with all our classes defined as "abstract" and when system starts (Web API service) framework can understand if a specific class has been inherited by customer or not. In case of not customized entities (most of them) the framework creates for each of them a concrete class (simple inheritance) and "Emit" it.
Of course our base logic doesn't know about concrete classes definition and always works using abstract classes. Customer can expand our base logic and use their own classes for their customizations..

and EF works in this scenario and, in this scenario, having always instance of the concrete class makes sense..
Thanks in advance

**Comment by ajcvickers** [comment]
@Luigi6821 I'm very surprised that EF6 allowed creation of a DbSet for an unmapped type. I will do some investigation into how that worked--would it be possible for you to spin up a simple console application showing how this is working with EF6?

**Comment by Luigi6821** [comment]
@ajcvickers 
Please find attached a running example (both EF and EF Core NET 5 console app)
EF Core doesn't work :-)
[ConsoleAppEFTest.zip](https://github.com/dotnet/efcore/files/5671369/ConsoleAppEFTest.zip)

**Comment by smitpatel** [comment]
This seems like TPC mapping.

**Comment by ajcvickers** [comment]
@Luigi6821 Thanks for the additional info. /cc @AndriySvyryd

**Comment by Luigi6821** [comment]
@smitpatel 
> This seems like TPC mapping.
Yes but mapped to same table.

**Comment by ajcvickers** [comment]
@Luigi6821 Thanks for the additional info. Putting this on the backlog to consider supporting this pattern in the future.

**Comment by Luigi6821** [comment]
> @Luigi6821 Thanks for the additional info. Putting this on the backlog to consider supporting this pattern in the future.

@ajcvickers  Thanks

**Comment by Xriuk** [comment]
I don't know if this is related but abstract base classes would be very handy if the base class itself is not needed to be created but only allow to access derived entities:
```csharp
// Entities
public abstract class Animal{
  public int Id {get; set;}

  [...]
}

public class Cat : Animal{
  [...]
}

public class Dog : Animal{
  [...]
}


// DbContext
public DbSet<Animal> Animals {get; set;}
public DbSet<Cat> Cats {get; set;}
public DbSet<Dog> Dogs {get; set;}
```

The Animals DbSet would retrieve all the animals and instantiate them as their correct concrete classes based on the discriminator, which would have only 2 values (Cat and Dog).

**Comment by ajcvickers** [comment]
@Xriuk This is normal TPH mapping. If this isn't working, then please open a new issue and attach a small, runnable project or post a small, runnable code listing that reproduces what you are seeing so that we can investigate.
